src/builders/rollup/plugins/cjs.ts

```diff
@@ -14,24 +14,83 @@ export function cjsPlugin(_opts?: any): Plugin {
   } as Plugin;
 }
 
-// Ported from https://github.com/egoist/tsup/blob/cd03e1e00ec2bd6676ae1837cbc7e618ab6a2362/src/rollup.ts#L92-L109
 export function fixCJSExportTypePlugin(): Plugin {
+  const regexp = /export\s*\{([^}]*)\}/;
+  const defaultExportRegexp = /\s*as\s+default\s*/;
+  const typeExportRegexp = /\s*type\s+/;
   return {
     name: "unbuild-fix-cjs-export-type",
-    renderChunk(code, info, opts) {
+    renderChunk(code, info) {
       if (
         info.type !== "chunk" ||
-        !info.fileName.endsWith(".d.cts") ||
+        !(
+          info.fileName.endsWith(".d.ts") || info.fileName.endsWith(".d.cts")
+        ) ||
         !info.isEntry ||
-        info.exports?.length !== 1 ||
-        info.exports[0] !== "default"
+        !info.exports?.length ||
+        !info.exports.includes("default")
       ) {
         return;
       }
 
+      const defaultExport = findExports(code).find((e) =>
+        e.names.includes("default"),
+      );
+
+      if (!defaultExport) {
+        return;
+      }
+
+      const match = defaultExport.code.match(regexp);
+      if (!match?.length) {
+        return;
+      }
+
+      let defaultAlias: string | undefined;
+      const exportsEntries: string[] = [];
+      for (const exp of match[1].split(",").map((e) => e.trim())) {
+        const m = exp.match(defaultExportRegexp);
+        if (m) {
+          defaultAlias = exp.replace(m[0], "");
+        } else {
+          exportsEntries.push(exp);
+        }
+      }
+
+      if (!defaultAlias) {
+        return;
+      }
+
+      let exportStatement = exportsEntries.length > 0 ? undefined : "";
+
+      // replace export { type A, type B, type ... } with export type { A, B, ... }
+      // that's, if all remaining exports are type exports, replace export {} with export type {}
+      if (exportStatement === undefined) {
+        let someExternalExport = false;
+        const allRemainingExports = exportsEntries.map((exp) => {
+          if (someExternalExport) {
+            return [exp, ""] as const;
+          }
+          if (!info.imports.includes(exp)) {
+            const m = exp.match(typeExportRegexp);
+            if (m) {
+              const name = exp.replace(m[0], "").trim();
+              if (!info.imports.includes(name)) {
+                return [exp, name] as const;
+              }
+            }
+          }
+          someExternalExport = true;
+          return [exp, ""] as const;
+        });
+        exportStatement = someExternalExport
+          ? `\nexport { ${allRemainingExports.map(([e, _]) => e).join(", ")} }`
+          : `\nexport type { ${allRemainingExports.map(([_, t]) => t).join(", ")} }`;
+      }
+
       return code.replace(
-        /(?<=(?<=[;}]|^)\s*export\s*){\s*([\w$]+)\s*as\s+default\s*}/,
-        `= $1`,
+        defaultExport.code,
+        `export = ${defaultAlias}${exportStatement}`,
```

Comment:
Looks like we need to add missing `;` after default export (shouldn't be a problem)